core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -531,7 +538,32 @@ private ExternalSearcherManager createSearcherManager(SearchFactory externalSear
         try {
             try {
                 final DirectoryReader directoryReader = ElasticsearchDirectoryReader.wrap(DirectoryReader.open(indexWriter), shardId);
-                internalSearcherManager = new SearcherManager(directoryReader, new SearcherFactory());
+                internalSearcherManager = new SearcherManager(directoryReader, new SearcherFactory() {
+                    @Override
+                    public IndexSearcher newSearcher(IndexReader reader, IndexReader previousReader) throws IOException {
+                        if (breakerService != null) {
+                            final CircuitBreaker breaker = breakerService.getBreaker(CircuitBreaker.ACCOUNTING);
+
+                            // get all new segment readers and account for their ram usage
+                            Set<SegmentReader> readers = reader.leaves().stream()
+                                    .map(c -> Lucene.segmentReader(c.reader())).collect(Collectors.toSet());
+                            Set<SegmentReader> prevReaders =
+                                    previousReader == null ? Collections.emptySet() : previousReader.leaves().stream()
+                                    .map(c -> Lucene.segmentReader(c.reader()))
+                                    .collect(Collectors.toSet());
+                            readers.removeAll(prevReaders);
+                            long totalNewRamBytesUsed = 0;
+                            for (SegmentReader r : readers) {
+                                final long ramBytesUsed = r.ramBytesUsed();
+                                totalNewRamBytesUsed += ramBytesUsed;
```

Comment:
I think you should add to the breaker as you go since if you run into an exception the close listeners might decrement once they are closed and then your accounting is messed up?

Reply:
Yes that's a good point, much better to do it that way.